File: day1/day1.py

import regex
import re
from pathlib import Path

import inflect

from aoc_utils import timing

numbers = []
with  open(Path(__file__).parent / 'input.txt', 'r') as fin: 
    for line in fin.readlines():
        vals = re.findall(r'\d', line)
        numbers.append(int(vals[0] + vals[-1]))
        
print(sum(numbers))


# Part 2 
numbers = []
mapping = {str(val): str(val) for val in range(0,10)}
words_mapping = {inflect.engine().number_to_words(val): str(val) for val in range(0, 10)}
mapping = mapping | words_mapping
pattern = regex.compile(r'\d|{}'.format("|".join(words_mapping)))
with timing():
    with  open(Path(__file__).parent / 'input.txt', 'r') as fin: 
        for line in fin.readlines():
            vals = pattern.findall(line, overlapped=True)
            numbers.append(int(mapping[vals[0]] + mapping[vals[-1]]))
    print(sum(numbers))

# part 2 without third party lib

def find_first_and_last(pattern: re.Pattern, string: str) -> tuple[str, str]:
    match = pattern.search(string)
    first = last = match.group()
    string = string[match.start() + 1:]
    while (match := pattern.search(string)):
        last = match.group()
        string = string[match.start() + 1:]
    return first, last

numbers = []
pattern = re.compile(r'\d|{}'.format("|".join(words_mapping)))
with timing():
    with  open(Path(__file__).parent / 'input.txt', 'r') as fin: 
        for line in fin.readlines():
            vals = find_first_and_last(pattern, line)
            numbers.append(int(mapping[vals[0]] + mapping[vals[-1]]))
print(sum(numbers))

File: day1/day1.jl
function get_sum(input::String)::Int
    re = r"[0-9]"
    lines = split(input, "\n")
    first_digit = map(x ->  x.match, filter(x -> ~isnothing(x), map(x-> match(re, x), lines)))
    second_digit = map(x -> x.match, filter(x -> ~isnothing(x), map(x-> match(re, reverse(x)), lines))) 
    return sum(map((a, b) -> parse(Int, a * b), first_digit, second_digit))
end

function match_all_with_overlapping(pattern::Regex, input::AbstractString)::Union{Tuple{String,String}, Nothing}
    m = match(pattern, input)
    first = last = m.match
    while ~isnothing(m)
        last = m.match
        m = match(pattern, input, m.offset + 1)
    end
    return first, last
end

function get_line_value(pattern::Regex, line::AbstractString, mapping:: Dict{String, String})::Int 
    a, b = match_all_with_overlapping(pattern, line)
    return parse(Int, get!(mapping, a, a) * get!(mapping, b, b))
end


function get_sum_with_strings(input::String)::Int
    mapping = Dict("one"=>"1", "two"=>"2", "three"=>"3", "four"=>"4", "five"=>"5",
                "six"=>"6", "seven"=>"7", "eight"=>"8", "nine"=>"9")
    pattern = Regex("[1-9]|" * join(keys(mapping), "|"))
    return sum(map(x -> get_line_value(pattern, x, mapping), 
                eachsplit(input, "\n", keepempty=false)))     
end


input = read(joinpath(@__DIR__, "input.txt"), String)
@time result = get_sum(input)
println(result)
input = read(joinpath(@__DIR__, "input.txt"), String)
@time result = get_sum_with_strings(input)
println(result)

